File: python/day9_part2_2024.py
class FileSegment:
    def __init__(self, id, start, end):
        self.id = id
        self.start = start
        self.end = end

def solve():
    with open("input.txt", "r") as f:
        line = f.readline().strip()

    # Create initial disk layout
    disk = []
    file_id = 0
    is_file = True
    for char in line:
        length = int(char)
        if is_file:
            disk.extend([str(file_id)] * length)
            file_id += 1
        else:
            disk.extend(['.'] * length)
        is_file = not is_file

    # Build file segments list
    files = []
    curr_id = None
    start = 0
    
    for i, val in enumerate(disk):
        if val == '.':
            curr_id = None
            continue
        
        file_id = int(val)
        if file_id != curr_id:
            curr_id = file_id
            start = i
            
        if i == len(disk) - 1 or (i + 1 < len(disk) and disk[i + 1] != val):
            files.append(FileSegment(file_id, start, i))

    # Process files from highest ID to lowest
    for file in reversed(files):
        file_len = file.end - file.start + 1
        leftmost_span = -1
        span_len = 0
        
        # Find leftmost suitable free space
        for i in range(file.start):
            if disk[i] == '.':
                if span_len == 0:
                    leftmost_span = i
                span_len += 1
                if span_len == file_len:
                    break
            else:
                span_len = 0
                leftmost_span = -1

        # Move file if we found suitable space
        if leftmost_span != -1 and span_len == file_len:
            # Clear old location
            for i in range(file.start, file.end + 1):
                disk[i] = '.'
            # Copy file to new location
            for i in range(file_len):
                disk[leftmost_span + i] = str(file.id)

    # Calculate checksum
    checksum = 0
    for i, val in enumerate(disk):
        if val != '.':
            checksum += i * int(val)
    print(checksum)

solve()

File: julia/day9_part2_2024.jl

function solve()
    line = strip(read("input.txt", String))
    disk = Char[]
    fileID = 0
    isFile = true
    for c in line
        length = parse(Int, string(c))
        if isFile
            append!(disk, fill(Char('0' + fileID), length))
            fileID += 1
        else
            append!(disk, fill('.', length))
        end
        isFile = !isFile
    end

    files = Tuple{Int, Int, Int}[]
    curID = -1
    start = 0
    for (i, b) in enumerate(disk)
        if b == '.'
            curID = -1
            continue
        end
        id = Int(b - '0')
        if id != curID
            curID = id
            start = i
        end
        if i == length(disk) || (i + 1 <= length(disk) && disk[i+1] - '0' != id)
            push!(files, (id, start, i))
        end
    end

    for i in length(files):-1:1
        f = files[i]
        fileLen = f[3] - f[2] + 1
        leftmostSpan = -1
        spanLen = 0
        for j in 1:f[2]-1
            if disk[j] == '.'
                if spanLen == 0
                    leftmostSpan = j
                end
                spanLen += 1
                if spanLen == fileLen
                    break
                end
            else
                spanLen = 0
                leftmostSpan = -1
            end
        end
        if leftmostSpan != -1 && spanLen == fileLen
            for x in f[2]:f[3]
                disk[x] = '.'
            end
            for x in 0:fileLen-1
                disk[leftmostSpan+x] = Char('0' + f[1])
            end
        end
    end

    checksum = 0
    for (i, b) in enumerate(disk)
        if b != '.'
            id = Int(b - '0')
            checksum += (i - 1) * id
        end
    end
    println(checksum)
end

solve()
